In [ ]:
from pprint import pprint

In [ ]:
!pip install datasets==3.6.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.5/491.5 kB 19.4 MB/s eta 0:00:00
  Attempting uninstall: datasets
    Found existing installation: datasets 4.0.0
    Uninstalling datasets-4.0.0:
      Successfully uninstalled datasets-4.0.0


In [ ]:
import datasets

#checking dataset version
print(datasets.__version__)

3.6.0


#Q.No - 1

Choose all the correct statements about the dataset at the following URL

https://huggingface.co/datasets/ai4bharat/naamapadam


Note: We strongly recommend using the appropriate functions in the datasets module to obtain the dataset’s information, rather than accessing that information directly from the hub.

In [ ]:
from datasets import get_dataset_config_names, load_dataset

print("Total number of  subsets ",len(get_dataset_config_names("ai4bharat/naamapadam")))

hindi = load_dataset("ai4bharat/naamapadam","hi")

print("Total number of training samples", len(hindi['train']))

print("Total number of subclasses", len(hindi['train'].features['ner_tags'].feature.names))



Total number of  subsets  11
Total number of training samples 985787
Total number of subclasses 7


#Q.No - 2
Common data

Download the Tamil sub-dataset from the following URL https://huggingface. co/datasets/ai4bharat/naamapadam, and store in a variable named “ds”.

Find the location of the cache directory (one way to determine the cache directory location is by using a method from the Dataset class; please refer to the documentation). Which of the following files are in the cached directory corresponding to the downloaded dataset?

In [ ]:
ds = load_dataset("ai4bharat/naamapadam","ta")

pprint(ds.cache_files)

{'test': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-test.arrow'}],
 'train': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-train.arrow'}],
 'validation': [{'filename': '/root/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-validation.arrow'}]}


#Q.No - 3
What is the disk space size of the downloaded dataset in MB?

In [ ]:
import os

size = 0

for file in ds.cache_files:
  for file_info in ds.cache_files[file]:
    size += os.path.getsize(file_info['filename'])

print("FIle size in MB",size/ (1024 * 1024))

FIle size in MB 179.28880310058594


#Q.No - 4
Enter the number of training samples in the dataset.

In [ ]:
print("Total number of  training samples",len(ds['train']))

Total number of  training samples 497882


#Q.No - 5
Create a new column named “num tokens” using the 'map' function.. Compute the number of tokens (words) in each sample and store the results in the newly created column. Reassign the modified dataset to the same variable “ds”. How many tokens (in millions) are there in the entire dataset?

In [ ]:
def num_tokens(rows):
  rows['num_tokens'] = len(rows['tokens'])
  return rows

In [ ]:
ds = ds.map(num_tokens)

Map:   0%|          | 0/497882 [00:00<?, ? examples/s]

Map:   0%|          | 0/758 [00:00<?, ? examples/s]

Map:   0%|          | 0/2795 [00:00<?, ? examples/s]

In [ ]:
total = 0
def total_tokens(rows):
  global total
  total += rows['num_tokens']

ds.map(total_tokens)
print("Total tokens in million",toal/10_00_000)

Map:   0%|          | 0/497882 [00:00<?, ? examples/s]

Map:   0%|          | 0/758 [00:00<?, ? examples/s]

Map:   0%|          | 0/2795 [00:00<?, ? examples/s]

NameError: name 'toal' is not defined

#Q.No - 6
The statement that the modified dataset increased the disk space requirement about two fold is

In [ ]:
size = 0
for file in ds.cache_files:
  for name in ds.cache_files[file]:
    size += os.path.getsize(name['filename'])
print("Total new file size in MB", size/(1024*1024) )

Total new file size in MB 183.1378173828125


In [ ]:
!ls -l ~/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc

total 739572
-rw-r--r-- 1 root root    326920 Jun 14 15:24 cache-992e0fcf59f3eeef.arrow
-rw-r--r-- 1 root root 190636448 Jun 14 15:24 cache-99dc831fb333eab6.arrow
-rw-r--r-- 1 root root 190636448 Jun 14 15:20 cache-9e92d4b337902d1d.arrow
-rw-r--r-- 1 root root 186629384 Jun 14 15:14 cache-ccc59e4cf514d66a.arrow
-rw-r--r-- 1 root root   1070552 Jun 14 15:24 cache-dd532199aff9666e.arrow
-rw-r--r-- 1 root root      1377 Jun 14 14:00 dataset_info.json
-rw-r--r-- 1 root root        75 Jun 14 14:00 LICENSE
-rw-r--r-- 1 root root    320672 Jun 14 14:00 naamapadam-test.arrow
-rw-r--r-- 1 root root 186629352 Jun 14 14:00 naamapadam-train.arrow
-rw-r--r-- 1 root root   1047912 Jun 14 14:00 naamapadam-validation.arrow
-rw------- 1 root root         0 Jun 14 15:09 tmpbr2gfulv
-rw------- 1 root root         0 Jun 14 15:10 tmpli9p0aqe
-rw------- 1 root root         0 Jun 14 15:10 tmpvpe2695r


Note : Though the file size doesn't gets modified here, there exits a slight variation, one must understand that, every time when we use the map funtion, the dataset gets duplicated and the file size gets doubled every time. So the new file size is,

Size = 179 + 183 = 362 (almost double)

#Q.No - 7

Concatenate all the samples across the splits in the following order: [train:test:validation]. Currently, each sample contains a list of words (tokens). Create a sentence by joining the individual words (tokens) in each sample using a single white space as a delimiter and store the re sulting sample in a new column named “text”. Create a new dataset by removing the columns “ner tags” and “tokens”. Store the new dataset in the same variable “ds”. Enter the total number of samples in “ds”. Use map function.

In [ ]:
print("Total number of data sets in the subset is equal to ",(len(ds['train'])+len(ds['test'])+len(ds['validation'])))

Total number of data sets in the subset is equal to  501435


In [ ]:
from datasets import concatenate_datasets

ds = concatenate_datasets([ds['train'],ds['test'],ds['validation']])

KeyError: "Column train not in the dataset. Current columns in the dataset: ['tokens', 'ner_tags', 'num_tokens']"

In [ ]:
def text(row):
  row['text'] = " ".join(row['tokens'])
  return row

In [ ]:
ds = ds.map(text)

Map:   0%|          | 0/501435 [00:00<?, ? examples/s]

In [ ]:
ds[0]

{'tokens': ['பைரவருக்கு',
  'தேய்பிறை',
  'அஷ்டமியில்',
  'விசேஷ',
  'அபிஷேக',
  'ஆராதனைகள்',
  'நடைபெறுகின்றன',
  '.'],
 'ner_tags': [1, 0, 0, 0, 0, 0, 0, 0],
 'num_tokens': 8,
 'text': 'பைரவருக்கு தேய்பிறை அஷ்டமியில் விசேஷ அபிஷேக ஆராதனைகள் நடைபெறுகின்றன .'}

In [ ]:
ds = ds.remove_columns(["ner_tags","tokens"])

In [ ]:
ds[0]

{'num_tokens': 8,
 'text': 'பைரவருக்கு தேய்பிறை அஷ்டமியில் விசேஷ அபிஷேக ஆராதனைகள் நடைபெறுகின்றன .'}

#Q.No - 9
Filter the dataset so that all the samples in the dataset should have at least six tokens (any symbol separated by a white space is considered a token). Enter the number of samples in the dataset after filtering. Enter the exact number. If the answer is 123456, enter it as 123456.

In [ ]:
ds = ds.filter(lambda x: x['num_tokens']>=6)

Filter:   0%|          | 0/501435 [00:00<?, ? examples/s]

In [ ]:
print("Dataset with atleast six tokens",len(ds))

370495


#Q.No - 10
Download all the splits of Tamil sub-dataset “inltkh.ta” of https://huggingface.co/datasets/ai4bharat/indic_glue .

Filter the dataset so that each sample contains at least six words (separated by a single white space).Then Interleave the resultant dataset with the above filtered dataset of naamapadam. Take 80% of samples from naamapadam and 20% from indic glue. Enter the number of samples after interleaving the datasets.
Note: Set the value of the seed argument to 42.

In [ ]:
tamil_new = load_dataset("ai4bharat/indic_glue","inltkh.ta",split="train+test+validation")

In [ ]:
tamil_new

Dataset({
    features: ['text', 'label'],
    num_rows: 6684
})

In [ ]:
tamil_new = tamil_new.filter(lambda x: len(x['text'].split(' '))>=6)

Filter:   0%|          | 0/6684 [00:00<?, ? examples/s]

In [ ]:
tamil_new

Dataset({
    features: ['text', 'label'],
    num_rows: 6428
})

In [ ]:
tamil_new = tamil_new.rename_column("label","num_tokens")

In [ ]:
tamil_new

Dataset({
    features: ['text', 'num_tokens'],
    num_rows: 6428
})

In [ ]:
ds

Dataset({
    features: ['num_tokens', 'text'],
    num_rows: 370495
})

In [ ]:
tamil_new = tamil_new.select_columns(["text"])
ds = ds.select_columns(["text"])

In [ ]:
tamil_new

Dataset({
    features: ['text'],
    num_rows: 6428
})

In [ ]:
ds

Dataset({
    features: ['text'],
    num_rows: 370495
})

In [ ]:
#Now we have to use the interleaving funtion to answer the above question.

from datasets import interleave_datasets

inter_data = interleave_datasets(
    datasets=[ds,tamil_new],
    probabilities=[0.8,0.2],
    seed=42
)

In [ ]:
print("Toal number of new samples after interleaving the dataset is equal to,",len(inter_data))

Toal number of new samples after interleaving the dataset is equal to, 32354
